src/main/java/com/google/sps/user/repository/impl/DatastoreUserRepository.java

```diff
@@ -66,15 +73,88 @@ private User getUserFromUserEntity(Entity userEntity) {
+    if (bookmarkedPlaceGuidesIdsList == null) {
+      bookmarkedPlaceGuidesIds = new HashSet<>();
+    } else {
+      bookmarkedPlaceGuidesIds = new HashSet<>(bookmarkedPlaceGuidesIdsList);
+    }
     Boolean publicPortfolio = (Boolean) userEntity.getProperty(PUBLIC_PORTFOLIO_PROPERTY);
     String selfIntroduction = (String) userEntity.getProperty(SELF_INTRODUCTION_PROPERTY);
     String imgKey = (String) userEntity.getProperty(IMG_KEY_PROPERTY);
     User.Builder newUserBuilder =
-            new User.Builder(id, email).setName(name).addSelfIntroduction(selfIntroduction).setPublicPortfolio(publicPortfolio).addImgKey(imgKey);
+        new User.Builder(id, email)
+        .setBookmarkedPlaceGuidesIds(bookmarkedPlaceGuidesIds)
+        .setName(name)
+        .addSelfIntroduction(selfIntroduction)
+        .setPublicPortfolio(publicPortfolio)
+        .addImgKey(imgKey);
     return newUserBuilder.build();
   }
 
+  @Override
   public boolean existingUser(String id) {
       return getUser(id) != null;
   }
+
+  @Override
+  public void bookmarkPlaceGuide(long placeGuideId, String userId) {
+    // Check if the corresponding placeGuide is in the database.
+    Key placeGuideEntityKey = 
+        KeyFactory.createKey(DatastorePlaceGuideRepository.ENTITY_KIND, placeGuideId);
+    if (placeGuideExists(placeGuideEntityKey)) {
+      User user = getUser(userId);
+      if (user != null) {
+        Set<Long> bookmarkedIds = user.getBookmarkedPlaceGuidesIds();
+        Set<Long> bookmarkedIdsCopy = new HashSet<>(bookmarkedIds);
+        bookmarkedIdsCopy.add(placeGuideId);
+
+      // Update and save user with the updated {@code bookmarkedPlaceGuides}.
+        saveUpdatedUser(user, bookmarkedIdsCopy);
+      } else {
+        throw new IllegalStateException("Non-existing user cannot bookmark a place guide!");
+      }
+    } else {
+      throw new IllegalStateException("Can't bookmark a non-existing place guide!");
+    }
+  }
+
+  private boolean placeGuideExists(Key placeGuideEntityKey) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    try {
+      datastore.get(placeGuideEntityKey);
+      return true;
+    } catch(EntityNotFoundException err) {
+      return false;
+    }
+  }
+
+  @Override
+  public void removeBookmarkedPlaceGuide(long placeGuideId, String userId) {
+    User user = getUser(userId);
+    if (user != null) {
+      Set<Long> bookmarkedPlaceGuidesIds = user.getBookmarkedPlaceGuidesIds();
+      Set<Long> bookmarkedPlaceGuidesIdsCopy = new HashSet<>(bookmarkedPlaceGuidesIds);
+      bookmarkedPlaceGuidesIdsCopy.remove(placeGuideId);
+      
+      // Update user with the new set.
+      saveUpdatedUser(user, bookmarkedPlaceGuidesIdsCopy);
+    } else {
+      throw new IllegalStateException("Cannot remove bookmarked place guides from non-existent user!");
+    }
+  }
+
+  private void saveUpdatedUser(User user, Set<Long> updatedBookmarkedPlaceGuidesIds) {
```

Comment:
Can we separate building the user and saving it to the database? Saving is already implemented, this function should handle building a new user with the updated bookmarkedPlaceGuideIds and returning the new User object only. Then you can save the the returned object, but these two tasks don't need to be done in the same function.